Feat: Manual Tracing

denrase opened the pull request:
# Overview

- Introduce manual tracing for stateful widgets
- Introduce manual tracing to sync of orgs and projects

# Motivation

We want to dog-food this feature and see which path we have to automatic tracing. So more suggestions, for example for app start or slow/frozen frames are more than welcome!

Closes https://github.com/getsentry/sentry-mobile/issues/219

# Assets

<img width="1819" alt="Screenshot 2021-11-04 at 15 06 41" src="https://user-images.githubusercontent.com/3984453/140328377-44a9917e-8b12-4d71-8d98-f902fc1c4ca7.png">

<img width="2259" alt="Screenshot 2021-11-04 at 15 06 25" src="https://user-images.githubusercontent.com/3984453/140328360-4f3502b8-497b-4c33-adee-677b0a79043a.png">

denrase commented:
@ueman Would be great to also have your thoughts on this. 🙇

marandaneto commented on lib/mixin/sentry_state_transaction_mixin.dart:
```diff
@@ -0,0 +1,21 @@
+      bindToScope: true,
+    );
+    WidgetsBinding.instance?.addPostFrameCallback((_) {
+      transaction?.finish();
```
you can use `SpanStatus.ok` also for all the other spans/transactions when there's no errors

denrase replied:
Is there no default?

marandaneto replied:
no, default is unknown by the server or UI

marandaneto commented on lib/redux/middleware/sentry_api_middleware.dart:
```diff
@@ -30,9 +32,20 @@ class SentryApiMiddleware extends MiddlewareClass<AppState> {
+              organizationSpan.status = SpanStatus.internalError();
+              rethrow;
+            } finally {
+              organizationSpan.finish();
```
same as https://github.com/getsentry/sentry-mobile/pull/220/files#r742880475

marandaneto commented on lib/redux/middleware/sentry_api_middleware.dart:
```diff
@@ -30,9 +32,20 @@ class SentryApiMiddleware extends MiddlewareClass<AppState> {
-                await api.organization(organization.slug);
-            individualOrganizations.add(individualOrganization);
+            final organizationSpan =
+                transaction.startChild('task', description: 'organization');
```
`task` is a very generic name, I'd rather do `organization` for operation name then and a more descriptive description

marandaneto commented on lib/redux/middleware/sentry_api_middleware.dart:
```diff
@@ -17,6 +17,8 @@ class SentryApiMiddleware extends MiddlewareClass<AppState> {
   dynamic call(Store<AppState> store, action, next) {
     if (action is FetchOrgsAndProjectsAction) {
       final thunkAction = (Store<AppState> store) async {
+        final transaction =
+            Sentry.startTransaction('FetchOrgsAndProjectsAction', 'task');
```
I'm not entirely sure a transaction makes sense here, I'd rather try to create a span out of the transaction bound to the scope

denrase replied:
Yeah, we might run into the case that first render is done before orgs and projects are synced. Also, I'm kinda confused as to what to choose for name and operation, the flutter docs don't really go into this in detail.

marandaneto replied:
we can discuss about that today over our call, but naming things is hard, do not bother about that right now

marandaneto commented on lib/mixin/sentry_state_transaction_mixin.dart:
```diff
@@ -0,0 +1,21 @@
+      'ui.load',
+      bindToScope: true,
+    );
+    WidgetsBinding.instance?.addPostFrameCallback((_) {
```
if `WidgetsBinding.instance` is null, `transaction` never finishes and leaks memory.

does `addPostFrameCallback` gets called for every screen? after rendering or what?

ueman replied:
At this point `WidgetsBinding.instance` can never be null.

denrase replied:
Does it make sense to check anyway?

ueman replied:
No, not really. There's already a PR in the Flutter repo somewhere which makes `WidgetsBinding.instance` non-nullable in a future version.

ueman commented:
@denrase You could additionally enable HTTP tracing with the `SentryHttpClient`.

ueman commented on lib/redux/middleware/sentry_api_middleware.dart:
```diff
@@ -45,27 +58,40 @@ class SentryApiMiddleware extends MiddlewareClass<AppState> {
                 rethrow;
               }
+            } finally {
+              projectIdsSpan.finish();
```
Is there a reason you never await `transaction.finish();` calls?

denrase replied:
No, just didn't notice it. Would like to have an option so the compiler would give me a hint.

ueman replied:
You can try to enable this lint: https://dart-lang.github.io/linter/lints/unawaited_futures.html

ueman commented on lib/screens/settings/settings_screen.dart:
```diff
@@ -210,7 +212,7 @@ class _SettingsState extends State<Settings> {
                           MaterialPageRoute(
                               fullscreenDialog: true,
-                              builder: (context) => SentryFlutterScreen()),
+                              builder: (context) => SentrySDKDebugScreen()),
```
Not really relevant to this PR, but `MaterialPageRoute` has a `settings` argument. If that's set, it'll be recorded as additional information in the navigation breadcrumbs.

marandaneto commented on lib/mixin/sentry_state_transaction_mixin.dart:
```diff
@@ -0,0 +1,21 @@
+      bindToScope: true,
+    );
+    WidgetsBinding.instance?.addPostFrameCallback((_) {
+      transaction?.finish(status: SpanStatus.ok());
```
See how we do to set status here https://github.com/getsentry/sentry-java/blob/main/sentry-android-core/src/main/java/io/sentry/android/core/ActivityLifecycleIntegration.java#L218-L223

denrase commented:
@marandaneto 

- Http tracking does work. I also changed the time when we are staring the org/projects sync for the main page, so that it is part of that transaction. However, first render of the widget is done before the http requests are finished, so the transaction is cut off before everything is ready.

<img width="1815" alt="Screenshot 2021-11-10 at 16 11 41" src="https://user-images.githubusercontent.com/3984453/141139951-bb8c349c-74e7-4044-b999-6d0f0ac8e247.png">

- Getting the name of the widgets state with `runtimeType.toString()` also works as expected, but I have also introduced a method to provide a custom name.

- Also went through flutter docs once more. I don't think we have much else to go by than the post frame callback for stateful widgets. Couldn't even find if post frame callbacks frame of reference is certainly associated with the rendering of the current widget, but it's how this API is used everywhere where I have looked, so I'm assuming this is correct.

- There is other API to get durations for frame rendering and rasterisation, so we could observe the performance of the whole application. But it looks like this is not specifically to a widget.

ueman commented:
> * Http tracking does work. I also changed the time when we are staring the org/projects sync for the main page, so that it is part of that transaction. However, first render of the widget is done before the http requests are finished, so the transaction is cut off before everything is ready.

It only works if a transaction is bound to the scope.

> * Getting the name of the widgets state with `runtimeType.toString()` also works as expected, but I have also introduced a method to provide a custom name.

I don't think using `runtimeType.toString()` is viable solution because the class names can be obfuscated. You wouldn't have any way to know what the transaction belongs to. Well, except manually de-obfuscating it without the help of Sentry or the the de-obfuscate tool.

> * Also went through flutter docs once more. I don't think we have much else to go by than the post frame callback for stateful widgets. Couldn't even find if post frame callbacks frame of reference is certainly associated with the rendering of the current widget, but it's how this API is used everywhere where I have looked, so I'm assuming this is correct.

The callback is for each frame, so for all widgets which are rendered in that time. So if there are a lot of widgets which use the mixin, you'll have a bunch of transactions which start at nearly the same time and end at the same time. I could imagine that could be a problem. You'll certainly have to really think about when to use it.

I'm not really concerned with widget build/render times because in my experience that's pretty much never the issue. Instead, the issue is mostly in expensive business logic. Dart is single threaded, so business logic can make a frame take a long time. That's why I prefer the `idle transaction` solution with transaction starting with a navigation event. This is also much more integrated out of the box with the `NavigationObserver`. If you're also adding frame metrics to the transaction, you also know when there's something wrong.

> * There is other API to get durations for frame rendering and rasterisation, so we could observe the performance of the whole application. But it looks like this is not specifically to a widget.

Yep, as seen [here](https://github.com/getsentry/sentry-dart/pull/519).

denrase commented:
@ueman Very good feedback, thank you!

- The transaction is bound to the scope, it's wehre we attach the child span from. Issue seems to be that we finish after first render, but child spans run for a longer time? I'm still missing context as to how this is done on other platforms, especially if child spans and http request take longer than the original transaction associated with the screen.
- In most cases rendering would happen in child widgets of that particular screen. The mixing needs to be added manually, so you'd only add it to you "main" navigation screens. But yeah, it's not perfect of course and we don't have render times of individual widgets available to us at runtime, unfortunately.
- Good point with the runtime type & obfuscation, this would not be feasible for the SDK, but it's fine here for testing / dog fooding. 👍 

So yeah, all in all, we are not sure if this approach here would provide enough value for users, that's also why we are adding it here and not in the SDK, to see some real-world data and take it from there.

marandaneto reviewed (commented):
left a few comments, they were pending so maybe they were resolved already, forgot to submit review, sorry
